wicketd/src/update_planner.rs

```diff
@@ -0,0 +1,489 @@
+
+    async fn reset_sp(&self) -> anyhow::Result<()> {
+        self.mgs_client
+            .sp_reset(self.sp.type_, self.sp.slot)
+            .await
+            .context("failed to reset SP")
+            .map(|res| res.into_inner())
+    }
+
+    async fn poll_for_component_update_completion(
+        &self,
+        artifact: &ArtifactId,
+        update_id: Uuid,
+        component: &str,
+    ) -> anyhow::Result<()> {
+        // How often we poll MGS for the progress of an update once it starts.
+        const STATUS_POLL_FREQ: Duration = Duration::from_millis(300);
+
+        loop {
+            let status = self
+                .mgs_client
+                .sp_component_update_status(
+                    self.sp.type_,
+                    self.sp.slot,
+                    component,
+                )
+                .await?
+                .into_inner();
+
+            match status {
+                SpUpdateStatus::None => {
+                    bail!("SP no longer processing update (did it reset?")
+                }
+                SpUpdateStatus::Preparing { id, progress } => {
+                    ensure!(id == update_id, "SP processing different update");
+                    self.set_current_update_state(
+                        UpdateStateKind::PreparingForArtifact {
+                            artifact: artifact.clone(),
+                            progress,
+                        },
+                    );
+                }
+                SpUpdateStatus::InProgress {
+                    bytes_received,
+                    id,
+                    total_bytes,
+                } => {
+                    ensure!(id == update_id, "SP processing different update");
+                    self.set_current_update_state(
+                        UpdateStateKind::ArtifactUpdateProgress {
+                            bytes_received: bytes_received.into(),
+                            total_bytes: total_bytes.into(),
+                        },
+                    );
+                }
+                SpUpdateStatus::Complete { id } => {
+                    ensure!(id == update_id, "SP processing different update");
+                    return Ok(());
+                }
+                SpUpdateStatus::Aborted { id } => {
+                    ensure!(id == update_id, "SP processing different update");
+                    bail!("update aborted");
+                }
+                SpUpdateStatus::Failed { code, id } => {
+                    ensure!(id == update_id, "SP processing different update");
+                    bail!("update failed (error code {code})");
+                }
+            }
+
+            tokio::time::sleep(STATUS_POLL_FREQ).await;
+        }
+    }
+}
+
+// Helper function to convert a `BufList` into a `Vec<u8>` for use with
+// `gateway_client::Client`.
+//
+// TODO-performance Can we pass `BufList`s directly (or wrapped) to
+// `gateway_client::Client` endpoints without copying the data?
```

Comment:
Definitely should work if we turn the RPC call into an `UntypedBody` and using `wrap_stream`: https://docs.rs/reqwest/latest/reqwest/struct.Body.html#method.wrap_stream

Reply:
Done in [47d1d1d](https://github.com/oxidecomputer/omicron/pull/2375/commits/47d1d1d51c466cf8d2f7d64fc0f1fb2f08002ed2)